File: sunflower-core/src/main/java/org/tdf/sunflower/pool/TransactionPoolImpl.java
package org.tdf.sunflower.pool;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import lombok.AllArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.tdf.common.event.EventBus;
import org.tdf.common.store.Store;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.ByteUtil;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.ApplicationConstants;
import org.tdf.sunflower.TransactionPoolConfig;
import org.tdf.sunflower.events.NewBestBlock;
import org.tdf.sunflower.events.NewTransactionsCollected;
import org.tdf.sunflower.facade.ConsensusEngine;
import org.tdf.sunflower.facade.PendingTransactionValidator;
import org.tdf.sunflower.facade.SunflowerRepository;
import org.tdf.sunflower.facade.TransactionPool;
import org.tdf.sunflower.state.Account;
import org.tdf.sunflower.state.StateTrie;
import org.tdf.sunflower.types.*;
import org.tdf.sunflower.vm.Backend;
import org.tdf.sunflower.vm.CallData;
import org.tdf.sunflower.vm.VMExecutor;

import java.util.*;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.function.Predicate;
import java.util.stream.Collectors;

@Slf4j(topic = "txPool")
@Component
public class TransactionPoolImpl implements TransactionPool {
    private final EventBus eventBus;

    // waited
    private final TreeSet<TransactionInfo> cache;
    // hash -> info
    private final Map<HexBytes, TransactionInfo> mCache;

    private final ScheduledExecutorService poolExecutor;
    private final TransactionPoolConfig config;
    private final SunflowerRepository repository;


    private final ReadWriteLock cacheLock = new ReentrantReadWriteLock();
    // dropped transactions
    private final Cache<HexBytes, Transaction> dropped;


    private PendingTransactionValidator validator;

    // pending transaction
    private Header parentHeader;
    private StateTrie<HexBytes, Account> trie;
    private List<Transaction> pending;
    private List<TransactionReceipt> pendingReceipts;
    private HexBytes currentRoot;

    public void reset(Header best) {
        this.cacheLock.writeLock().lock();
        try {
            this.parentHeader = best;
            this.currentRoot = best.getHash();
            this.pending = new ArrayList<>();
            this.pendingReceipts = new ArrayList<>();
        } finally {
            this.cacheLock.writeLock().unlock();
        }
    }

    public TransactionPoolImpl(
        EventBus eventBus,
        TransactionPoolConfig config,
        SunflowerRepository repository
    ) {
        this.eventBus = eventBus;
        eventBus.subscribe(NewBestBlock.class, this::onNewBestBlock);
        cache = new TreeSet<>();
        this.config = config;


        poolExecutor = Executors.newSingleThreadScheduledExecutor();
        poolExecutor.scheduleWithFixedDelay(this::clear, 0, config.getExpiredIn(), TimeUnit.SECONDS);


        dropped = CacheBuilder.newBuilder()
            .expireAfterWrite(config.getExpiredIn(), TimeUnit.SECONDS)
            .build();


        this.repository = repository;
        this.mCache = new HashMap<>();

    }

    public void setEngine(ConsensusEngine engine) {
        this.validator = engine.getValidator();
        this.trie = engine.getAccountTrie();
        Header best = repository.getBestHeader();
        this.parentHeader = best;
        this.currentRoot = best.getStateRoot();
        this.pending = new ArrayList<>();
        this.pendingReceipts = new ArrayList<>();
    }

    private void clearPending() {
        this.pending = new ArrayList<>();
        this.pendingReceipts = new ArrayList<>();
        this.currentRoot = null;
        this.parentHeader = null;
    }

    @SneakyThrows
    private void clear() {
        long now = System.currentTimeMillis();
        if (!this.cacheLock.writeLock().tryLock(config.getLockTimeout(), TimeUnit.SECONDS)) {
            return;
        }
        try {
            Predicate<TransactionInfo> lambda =
                info -> {
                    boolean remove = (now - info.receivedAt) / 1000 > config.getExpiredIn();
                    if (remove) {
                        dropped.put(info.tx.getHashHex(), info.tx);
                    }
                    return remove;
                };
            cache.removeIf(lambda);
            mCache.values().removeIf(lambda);
        } finally {
            this.cacheLock.writeLock().unlock();
        }
    }

    @Override
    @SneakyThrows
    public Map<HexBytes, String> collect(Collection<? extends Transaction> transactions) {
        Map<HexBytes, String> errors = new HashMap<>();
        this.cacheLock.writeLock().lock();
        try {
            List<Transaction> newCollected = new ArrayList<>(transactions.size());
            for (Transaction transaction : transactions) {
                if (transaction.getGasPriceAsU256().compareTo(Uint256.of(ApplicationConstants.VM_GAS_PRICE)) < 0)
                    throw new RuntimeException("transaction pool: gas price of tx less than vm gas price " + ApplicationConstants.VM_GAS_PRICE);

                TransactionInfo info = new TransactionInfo(System.currentTimeMillis(), transaction);
                if (cache.contains(info) || dropped.asMap().containsKey(transaction.getHashHex()))
                    continue;
                try {
                    transaction.verify();
                } catch (Exception e) {
                    log.error(e.getMessage());
                    continue;
                }
                ValidateResult res = validator.validate(parentHeader, transaction);
                if (res.isSuccess()) {
                    cache.add(info);
                    mCache.put(info.tx.getHashHex(), info);
                    newCollected.add(transaction);
                } else {
                    log.error(res.getReason());
                    errors.put(transaction.getHashHex(), res.getReason());
                }
            }
            if (!errors.isEmpty())
                newCollected = Collections.emptyList();

            if (!newCollected.isEmpty())
                eventBus.publish(new NewTransactionsCollected(newCollected));

            this.execute(errors);
        } finally {
            this.cacheLock.writeLock().unlock();
        }
        return errors;
    }

    @SneakyThrows
    private void execute(Map<HexBytes, String> errors) {
        Iterator<TransactionInfo> it = cache.iterator();
        if(this.parentHeader == null)
            return;

        Backend current = trie.createBackend(parentHeader, currentRoot, null, false);

        while (it.hasNext()) {
            Transaction t = it.next().tx;
            long prevNonce = current.getNonce(t.getSenderHex());

            if (t.getNonceAsLong() < prevNonce) {
                it.remove();
                errors.put(t.getHashHex(), "nonce is too small");
                mCache.remove(t.getHashHex());
                dropped.put(t.getHashHex(), t);
                continue;
            }

            if (t.getNonceAsLong() != prevNonce) {
                continue;
            }

            // try to execute

            try {
                Backend child = current.createChild();
                CallData callData = CallData.fromTransaction(t, false);
                VMExecutor vmExecutor = new VMExecutor(child, callData);
                VMResult res = vmExecutor.execute();

                // execute successfully
                long currentGas =
                    pendingReceipts.size() > 0 ?
                        ByteUtil.byteArrayToLong(pendingReceipts.get(pendingReceipts.size() - 1).getCumulativeGas()) : 0;

                TransactionReceipt receipt = new TransactionReceipt(
                    current.getTrieRoot().getBytes(),
                    ByteUtil.longToBytesNoLeadZeroes(currentGas + res.getGasUsed()),
                    new Bloom(),
                    Collections.emptyList()
                );

                receipt.setGasUsed(res.getGasUsed());
                receipt.setExecutionResult(res.getExecutionResult());
                receipt.setTransaction(t);
                pending.add(t);
                pendingReceipts.add(receipt);
                this.currentRoot = child.merge();
            } catch (Exception e) {
                errors.put(t.getHashHex(), e.getMessage());
            } finally {
                it.remove();
                mCache.remove(t.getHashHex());
            }
        }

    }

    @SneakyThrows
    public PendingData pop(Header parentHeader) {
        this.cacheLock.writeLock().lock();
        try {
            if(this.parentHeader != null && !parentHeader.getHash().equals(this.parentHeader.getHash())) {
                this.clearPending();
                log.warn("parent header is not equal, drop pending");
            }
            boolean afterClean = this.parentHeader == null;
            PendingData r =
                new PendingData(
                    afterClean ? new ArrayList<>() : this.pending,
                    afterClean ? new ArrayList<>() : this.pendingReceipts,
                    afterClean ? parentHeader.getStateRoot() : this.currentRoot);
            return r;
        } finally {
            this.cacheLock.writeLock().unlock();
        }
    }

    @SneakyThrows
    public void onNewBestBlock(NewBestBlock event) {
        this.cacheLock.writeLock().tryLock();
        try {
            if(event.getBlock().getStateRoot().equals(this.parentHeader.getStateRoot()))
                return;

            this.parentHeader = event.getBlock().getHeader();
            this.currentRoot = event.getBlock().getStateRoot();
            this.pending = new ArrayList<>();
            this.pendingReceipts = new ArrayList<>();
        } finally {
            this.cacheLock.writeLock().unlock();
        }
    }


    @AllArgsConstructor
    static class TransactionInfo implements Comparable<TransactionInfo> {
        private final long receivedAt;
        private final Transaction tx;

        @Override
        public int compareTo(TransactionInfo o) {
            int cmp = tx.getSenderHex().compareTo(o.tx.getSenderHex());
            if (cmp != 0) return cmp;
            cmp = Long.compare(tx.getNonceAsLong(), o.tx.getNonceAsLong());
            if (cmp != 0) return cmp;
            cmp = -tx.getGasPriceAsU256().compareTo(o.tx.getGasPriceAsU256());
            if (cmp != 0) return cmp;
            return tx.getHashHex().compareTo(o.tx.getHashHex());
        }
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/pool/TransactionPoolImpl.kt
package org.tdf.sunflower.pool

import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.tdf.common.event.EventBus
import org.tdf.common.util.FixedDelayScheduler
import org.tdf.common.util.HexBytes
import org.tdf.common.util.LogLock
import org.tdf.sunflower.AppConfig
import org.tdf.sunflower.TransactionPoolConfig
import org.tdf.sunflower.events.NewBestBlock
import org.tdf.sunflower.events.NewTransactionsCollected
import org.tdf.sunflower.facade.*
import org.tdf.sunflower.state.Account
import org.tdf.sunflower.state.AddrUtil
import org.tdf.sunflower.state.StateTrie
import org.tdf.sunflower.types.*
import org.tdf.sunflower.vm.Backend
import org.tdf.sunflower.vm.CallContext
import org.tdf.sunflower.vm.CallData
import org.tdf.sunflower.vm.VMExecutor
import java.math.BigInteger
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReadWriteLock
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.withLock
import kotlin.math.min

@Component
class TransactionPoolImpl(
    private val eventBus: EventBus,
    private val config: TransactionPoolConfig,
    private val repo: RepositoryService,
    private val trie: StateTrie<HexBytes, Account>,
    private val appCfg: AppConfig,
    private val conCfg: ConsensusConfig
) : TransactionPool {
    companion object {
        val log: Logger = LoggerFactory.getLogger("txPool")
    }

    // waiting, sorted by gasPrice
    private val waiting: TreeSet<WaitingInfo> = TreeSet()

    private val lock: ReadWriteLock = ReentrantReadWriteLock()
    private val readLock = LogLock(lock.readLock(), "tp-r")
    private val writeLock = LogLock(lock.writeLock(), "tp-w")

    // dropped transactions
    val cache: Cache<HexBytes, Dropped> =
        CacheBuilder.newBuilder()
            .expireAfterWrite(15, TimeUnit.SECONDS)
            .build()

    override val dropped: MutableMap<HexBytes, Dropped> get() = cache.asMap()

    // pending records
    // waiting -> pending -> included
    private var pendingRec: PendingRec = PendingRec()

    private val clearScheduler = FixedDelayScheduler("txPool-clear", config.expiredIn)

    private lateinit var gasLimit: BigInteger


    fun init() {
        repo.reader.use {
            pendingRec.reset(it.bestHeader)
            gasLimit = it.genesis.gasLimit.toBigInteger()
        }
    }


    /**
     * clear waiting transactions
     */
    private fun clear() {
        val now = System.currentTimeMillis()
        if (!writeLock.tryLock(config.lockTimeout, TimeUnit.SECONDS)) {
            return
        }
        try {
            waiting.removeIf {
                val remove = (now - it.receivedAt) / 1000 > config.expiredIn
                if (remove) {
                    dropped[it.tx.hash] = Dropped(it.tx, "transaction timeout")
                }
                remove
            }
        } finally {
            writeLock.unlock()
        }
    }


    override fun collect(
        rd: RepositoryReader,
        transactions: Collection<Transaction>,
        source: String
    ): Map<HexBytes, String> {
        val errors: MutableMap<HexBytes, String> = mutableMapOf()

        writeLock.withLock {
            val newCollected: MutableList<Transaction> = mutableListOf()
            for (tx in transactions) {
                log.debug(
                    "new tx {} received from {} rpc sender = {}, nonce = {}",
                    tx.hash,
                    source,
                    tx.sender,
                    tx.nonce
                )
                if (tx.gasPrice < appCfg.vmGasPrice) {
                    errors[tx.hash] = "transaction pool: gas price of tx less than vm gas price ${appCfg.vmGasPrice}"
                    continue
                }

                if (rd.containsTransaction(tx.hash))
                    continue

                val drop = dropped[tx.hash]
                if (drop != null) {
                    errors[tx.hash] = drop.err
                    continue
                }

                val info = WaitingInfo(System.currentTimeMillis(), tx)
                if (pendingRec.pending.contains(info) || waiting.contains(info)) {
                    continue
                }

                try {
                    tx.validate()
                } catch (e: Exception) {
                    errors[tx.hash] = e.message ?: "validate transaction failed"
                    continue
                }
                if (tx.vrs == null || !tx.verifySig) {
                    errors[tx.hash] = "validate signature failed"
                    continue
                }

                if (tx.chainId != appCfg.chainId) {
                    errors[tx.hash] = "invalid chainId ${tx.chainId}"
                    continue
                }

                log.debug("add tx {} from {} to waiting sender = {} nonce = {}", tx.hash, source, tx.sender, tx.nonce)
                waiting.add(info)
                newCollected.add(tx)
            }

            pendingRec.execute(rd, waiting.iterator())

            transactions.forEach { t ->
                this.dropped[t.hash]?.let { errors[t.hash] = it.second }
            }
            if (newCollected.isNotEmpty())
                eventBus.publish(NewTransactionsCollected(newCollected))
            return errors
        }
    }

    internal data class PendingRec(
        var backend: Backend? = null,
        var gas: Long = 0,
        var timestamp: Long = 0,
        val pending: MutableList<WaitingInfo> = mutableListOf(),
        val receipts: MutableList<TransactionReceipt> = mutableListOf()
    )

    private fun PendingRec.clear() {
        backend = null
        pending.clear()
        receipts.clear()
        timestamp = 0
        gas = 0
    }

    private fun PendingRec.rollback(best: Header, timestamp: Long? = null) {
        val ts = timestamp ?: if (best.height == 0L) {
            System.currentTimeMillis() / 1000
        } else {
            best.createdAt + conCfg.blockInterval
        }
        waiting.addAll(pending)
        pending.clear()
        receipts.clear()
        this.timestamp = ts
        gas = 0
        backend = trie.createBackend(best)
    }

    // reset pending record to a header
    private fun PendingRec.reset(best: Header) {
        pending.clear()
        receipts.clear()
        gas = 0
        backend = trie.createBackend(best)
        timestamp = if (best.height == 0L) {
            System.currentTimeMillis() / 1000
        } else {
            best.createdAt + conCfg.blockInterval
        }
    }

    // remove from waiting if success, keep unchanged if nonce too big, moved to drop if nonce too small or failed
    private fun PendingRec.execute(rd: RepositoryReader, waiting: MutableIterator<WaitingInfo>) {
        var backend = this.backend ?: return

        while (waiting.hasNext()) {
            val n = waiting.next()
            val t = n.tx
            val prevNonce = backend.getNonce(t.sender)

            if (t.nonce < prevNonce) {
                waiting.remove()
                dropped[t.hash] = Pair(t, "nonce is too small")
                continue
            }

            if (t.nonce != prevNonce) {
                continue
            }

            val blockGasLimit = gasLimit.longValueExact()

            try {
                val child = backend.createChild()
                val ctx = CallContext.fromTx(
                    tx = t,
                    timestamp = this.timestamp,
                    coinbase = conCfg.coinbase ?: AddrUtil.empty(),
                    blockHashMap = rd.createBlockHashMap(child.parentHash).toMap()
                )
                val callData = CallData.fromTx(t)
                val vmExecutor = VMExecutor.create(
                    rd,
                    child, ctx, callData,
                    min(t.gasLimit, blockGasLimit)
                )

                val res = vmExecutor.execute()

                if (res.gasUsed > blockGasLimit) {
                    waiting.remove()
                    dropped[t.hash] = Pair(t, "gas overflows block gas limit")
                    continue
                }

                // execute successfully
                // skip high gas transaction
                if (gas + res.gasUsed > blockGasLimit) {
                    continue
                }

                val receipt = TransactionReceipt(
                    cumulativeGas = gas + res.gasUsed,
                    logInfoList = res.logs,
                    gasUsed = res.gasUsed,
                    result = res.executionResult
                )

                this.pending.add(n)
                this.receipts.add(receipt)
                backend = child
                this.backend = backend
                this.gas += res.gasUsed
                log.debug(
                    "add tx {} to pending, ready to pack, sender = {}, nonce = {} remaining gas = {}",
                    n.tx.hash,
                    n.tx.sender,
                    n.tx.nonce,
                    blockGasLimit - gas
                )
                waiting.remove()
            } catch (e: Exception) {
                e.printStackTrace()
                log.info("dropped tx ${n.tx.hash}, execute failed, gas limit = ${n.tx.gasLimit}")
                dropped[t.hash] = Pair(t, e.message ?: "execute tx error")
                waiting.remove()
            }
        }
    }

    override fun pop(rd: RepositoryReader, parentHeader: Header, timestamp: Long): PendingData {
        writeLock.withLock {
            log.debug("pendings = {}, waitings = {}", this.pendingRec.pending.size, this.waiting.size)
            if (parentHeader.hash != this.pendingRec.backend?.parentHash || timestamp != pendingRec.timestamp) {
                pendingRec.rollback(parentHeader, timestamp)
                pendingRec.execute(rd, waiting.iterator())
            }

            val pending = pendingRec.pending.toList().map { it.tx }

            val receipts = pendingRec.receipts.toList()
            val cur = pendingRec.backend
            pendingRec.clear()
            log.debug("pop {} transactions to miner", pending.size)
            return PendingData(pending, receipts, cur)
        }
    }

    override fun reset(parent: Header) {
        writeLock.withLock {
            pendingRec.reset(parent)
        }
    }

    override fun current(): Backend? {
        readLock.withLock {
            return pendingRec.backend?.createChild()
        }
    }

    private fun onNewBestBlock(event: NewBestBlock) {
        if (!writeLock.tryLock()) {
            return
        }
        try {
            if (event.block.hash == this.pendingRec.backend?.parentHash) return
            pendingRec.rollback(event.block)
        } finally {
            writeLock.unlock()
        }
    }

    internal class WaitingInfo(val receivedAt: Long, val tx: Transaction) : Comparable<WaitingInfo> {
        override fun compareTo(other: WaitingInfo): Int {
            if (tx.hash == other.tx.hash)
                return 0
            var cmp = -tx.gasPrice.compareTo(other.tx.gasPrice)
            if (cmp != 0) return cmp
            cmp = tx.sender.compareTo(other.tx.sender)
            if (cmp != 0) return cmp
            cmp = tx.nonce.compareTo(other.tx.nonce)
            if (cmp != 0) return cmp
            return tx.hash.compareTo(other.tx.hash)
        }
    }

    init {
        eventBus.subscribe(NewBestBlock::class.java) { onNewBestBlock(it) }
        clearScheduler.delay {
            try {
                clear()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}